examples/vision/depth_estimation.py

```diff
@@ -0,0 +1,507 @@
+                - tf.image.ssim(
+                    target, pred, max_val=256, filter_size=7, k1=0.01 ** 2, k2=0.03 ** 2
+                )
+            )
+            # Point-wise depth
+            l1_loss = tf.reduce_mean(tf.abs(target - pred))
+
+            # Weights
+            w1 = 0.85
+            w2 = 0.1
+            w3 = 1.0
+
+            loss = (w1 * l_ssim) + w2 * l1_loss + (w3 * l_edges)
+
+        gradients = tape.gradient(loss, self.model.trainable_variables)
+        self.optimizer.apply_gradients(zip(gradients, self.model.trainable_variables))
+        self.loss_metric.update_state(loss)
+        return {
+            "Loss": self.loss_metric.result(),
+        }
+
+    def call(self, x):
+        pass
+
+
+"""
+## Model training
+"""
+
+optimizer = tf.keras.optimizers.Adam(
+    learning_rate=config.LR,
+    amsgrad=True,
+)
+model = UnetModel(config.HEIGHT, config.WIDTH)
+DEM = DepthEstimationModel(model=model)
+DEM.compile(optimizer)
+
+train_loader = DataGenerator(
+    data=df[:600].reset_index(drop="true"),
+    batch_size=config.BATCH_SIZE,
+    dim=(config.HEIGHT, config.WIDTH),
+)
+DEM.fit(train_loader, epochs=config.EPOCHS)
+
+"""
+## Visualizing Model output.
+Visualizing model output over validation set.
+The first image is the RGB image, the second image is the ground truth depth-map image
+and the third one is the predicted depth-map image.
+"""
+
+test_loader = next(
+    iter(
+        DataGenerator(
+            data=df[601:].reset_index(drop="true"),
+            batch_size=6,
+            dim=(config.HEIGHT, config.WIDTH),
+        )
+    )
+)
+input, target = test_loader
+pred = model.predict(input)
+cmap = plt.cm.jet
+cmap.set_bad(color="black")
+fig, ax = plt.subplots(6, 3, figsize=(50, 50))
+for i in range(6):
+    ax[i, 0].imshow((input[i].squeeze()))
+    ax[i, 1].imshow((target[i].squeeze()), cmap=cmap)
+    ax[i, 2].imshow((pred[i].squeeze()), cmap=cmap)
+
+test_loader = next(
+    iter(
+        DataGenerator(
+            data=df[701:].reset_index(drop="true"),
+            batch_size=6,
+            dim=(config.HEIGHT, config.WIDTH),
+        )
+    )
+)
+input, target = test_loader
+pred = model.predict(input)
+cmap = plt.cm.jet
+cmap.set_bad(color="black")
+fig, ax = plt.subplots(6, 3, figsize=(50, 50))
+for i in range(6):
+    ax[i, 0].imshow((input[i].squeeze()))
+    ax[i, 1].imshow((target[i].squeeze()), cmap=cmap)
+    ax[i, 2].imshow((pred[i].squeeze()), cmap=cmap)
+
+"""
+## Scopes of Improvement
+
+1. From the research papers that I read while I was working on this topic the encode part
+of the unet was replaced with DenseNet, ResNet or other pre-trained model, which could be
+applied for better model predictions.
+
+2. Loss functions plays an immense role in solving this problem, and different paper
+explained different ways of developing the loss function out of which SSIM was common in
+all. so playing with the loss fuctions gives a huge scope of improvement in this case.
```

Comment:
comma before `so` instead of fullstop, and check spelling of functions

Reply:
updated